system/lib/wasmfs/wasmfs.cpp

```diff
@@ -9,25 +9,34 @@
 #include <emscripten/emscripten.h>
 #include <emscripten/html5.h>
 #include <stdlib.h>
+#include <string>
 #include <wasi/api.h>
 
 extern "C" {
 
-int emscripten_wasmfs_printbuffer(__wasi_fd_t fd, const uint8_t* ptr, __wasi_size_t len);
+std::string buffer;
```

Comment:
Could this be a `std::vector<char>`? That would avoid including code for string support.

Reply:
Would this also not bring in `libc++` as well if vector is part of stdlib?

Reply:
```
wasm-ld: error: /mnt/c/Users/nathe/Documents/Github/emscripten/cache/sysroot/lib/wasm32-emscripten/libwasmfs.a(wasmfs.o): undefined symbol: typeinfo for std::length_error
wasm-ld: error: /mnt/c/Users/nathe/Documents/Github/emscripten/cache/sysroot/lib/wasm32-emscripten/libwasmfs.a(wasmfs.o): undefined symbol: vtable for std::length_error
```

Reply:
`error: undefined symbol: _ZNKSt3__220__vector_base_commonILb1EE20__throw_length_errorEv (referenced by top-level compiled C/C++ code)`

Reply:
```
wasm-ld: error: emscripten/cache/sysroot/lib/wasm32-emscripten/libwasmfs.a(wasmfs.o): undefined symbol: operator delete(void*)
wasm-ld: error: emscripten/cache/sysroot/lib/wasm32-emscripten/libwasmfs.a(wasmfs.o): undefined symbol: std::__2::__vector_base_common<true>::__throw_length_error() const
wasm-ld: error: emscripten/cache/sysroot/lib/wasm32-emscripten/libwasmfs.a(wasmfs.o): undefined symbol: operator new(unsigned long)
```

Reply:
The `throw` symbol should vanish without exceptions, I think, so we should figure out what's going on there.

But `new/delete` do make sense. @sbc100 could we force libc++abi to be linked in all the time? When eventually this is turned on by default, we'd need that I think.

Reply:
I think this is happening because in the test suite we inject `-nostdlib++` (because by default its still included):

https://github.com/emscripten-core/emscripten/blob/aca7c128fd844d2b2b306aa2d5205a07f053bbe0/tests/common.py#L558

I'd be ok with only doing this when WASMFS was not in settings.

Reply:
I can do that. We can also look into building with cflag `'-fno-exceptions'`

Reply:
I'm not sure that will be needed.

Reply:
> The throw symbol should vanish without exceptions, I think, so we should figure out what's going on there.

Is there another way to resolve this symbol: `__throw_length_error` ?

I was just trying different things out locally

Reply:
Does simply building with `-nostdlib++` fix that?

Reply:
Yes, I didn't use both. I was just trying out Alon's suggestion to use `-fno-exceptions` to see if  the `throw` symbol warning would be resolved

Reply:
@sbc100 

> I'd be ok with only doing this when WASMFS was not in settings.

Would you be ok, though, if in the long term we needed to keep passing some C++ stdlibs to the linker, always? When WASMFS becomes the default, that will be required, if we allow C++ in WASMFS itself.

Reply:
Yes, that is a little unfortunate.   I was hoping we could so something like:

```
if settings.WASMFS:
  settings.DEFAULT_TO_CXX=1
```

But if that means pretty much all users will get `DEFAULT_TO_CXX`.. that is a little unfortunate.    Removing libc++/libstdc++ from C programs is more of a nice-to-have.. its really not that important.   If we can write WASMFS in C that would be good but it seems like that would slow us down quite lot, right?  Probably not worth it.

Reply:
Yes, there are tradeoffs here... I think writing WASMFS using modern C++ would be good (`unique_ptr`, etc. would avoid a lot of bugs). But it does mean that turning WASMFS on by default would mean `DEFAULT_TO_CXX` everywhere.

For completeness, we've also thrown around the idea of writing WASMFS in Rust or another language that would be better than C. But the build system integration work for that would be pretty large, sadly.

If you're ok with this then I think we can proceed here. But we should also look to see if we can avoid ``DEFAULT_TO_CXX`` somehow. I wonder if we can bundle the C++ requirements of WASMFS inside libwasmfs - those should be very few and very trivial, things like stubs for new/delete that just call malloc/free, etc. We'd avoid practically all C++ that uses stuff that is not in headers.